src/vm.rs

```diff
@@ -549,6 +566,15 @@ where
         let (_, r1cs_var) = self.attach_variable(var);
         Expression {
             terms: vec![(r1cs_var, Scalar::one())],
+            assignment: self.variable_assignment(var),
+        }
+    }
+
+    fn variable_assignment(&mut self, var: Variable) -> Option<u64> {
```

Comment:
as mentioned above, the flavor is a full-size `Scalar`, not u64, so we should not assume u64 anywhere outside specifically `qty` and range-proven values.

Reply:
Oh, i see. If the variable has u64 assignment, we don't want to type-erase it and then attempt to recover it... 🤔

Reply:
I guess since we have a generic `Variable` that has to enter spacesuit world, we need two kinds of witnesses for it: integer and scalar. Integer is u64, and scalar is Scalar.

Therefore, `CommitmentWitness` needs to be:

```rust
#[derive(Clone, Debug)]
pub struct CommitmentWitness {
    value: ScalarKind,
    blinding: Scalar,
}

#[derive(Clone, Debug)]
pub enum ScalarKind {
    Integer(u64),
    Scalar(Scalar)
}

...

impl Into<Scalar> for ScalarKind {
    fn into(self) -> Scalar {
        match(self) {
            ScalarKind::Integer(i) => i.into(),
            ScalarKind::Scalar(s) => s
        }
    }
}
```